tests/linearizability/model.go

```diff
@@ -104,50 +105,70 @@ func step(state EtcdState, request etcdRequest, response etcdResponse) (bool, Et
 	}
 }
 
-func initState(request etcdRequest, response etcdResponse) EtcdState {
+func initState(request EtcdRequest, response EtcdResponse) EtcdState {
 	state := EtcdState{
-		Key:          request.key,
-		LastRevision: response.revision,
-		FailedWrites: map[string]struct{}{},
+		Key:          request.Key,
+		LastRevision: response.Revision,
 	}
-	switch request.op {
+	switch request.Op {
 	case Get:
-		state.Value = response.getData
+		state.Value = response.GetData
 	case Put:
-		if response.err == nil {
-			state.Value = request.putData
+		if response.Err == nil {
+			state.Value = request.PutData
 		} else {
-			state.FailedWrites[request.putData] = struct{}{}
+			state.FailedWrite = &request
+			state.FailedCount += 1
 		}
 	default:
 		panic("Unknown operation")
 	}
 	return state
 }
 
-func stepGet(state EtcdState, request etcdRequest, response etcdResponse) (bool, EtcdState) {
-	if state.Value == response.getData && state.LastRevision <= response.revision {
+func stepGet(state EtcdState, request EtcdRequest, response EtcdResponse) (bool, EtcdState) {
+	if state.Value == response.GetData && state.LastRevision == response.Revision {
+		state.LastRevision = response.Revision
+		state.FailedWrite = nil
+		state.FailedCount = 0
 		return true, state
 	}
-	_, ok := state.FailedWrites[response.getData]
-	if ok && state.LastRevision < response.revision {
-		state.Value = response.getData
-		state.LastRevision = response.revision
-		delete(state.FailedWrites, response.getData)
-		return true, state
+	if state.FailedWrite != nil && state.LastRevision < response.Revision && state.LastRevision+state.FailedCount >= response.Revision {
```

Comment:
I'm wondering whether you can just keep an "expectedRevision" around instead of doing the mental math around failed requests and response revisions. Then this check and the one below would be fairly straight forward != checks.

Reply:
Decided that this is overkill for now. I will create a separate PR that will make this more readable.